File: packages/pdf-generator/android/src/main/java/io/capawesome/capacitorjs/plugins/pdfgenerator/classes/options/GenerateOptions.java
package io.capawesome.capacitorjs.plugins.pdfgenerator.classes.options;

import android.print.PrintAttributes;
import androidx.annotation.NonNull;
import com.getcapacitor.PluginCall;
import io.capawesome.capacitorjs.plugins.pdfgenerator.classes.CustomExceptions;
import java.util.UUID;

public class GenerateOptions {

    @NonNull
    private final String fileName;

    @NonNull
    private final PrintAttributes.MediaSize mediaSize;

    private final long timeout;

    public GenerateOptions(@NonNull PluginCall call) throws Exception {
        this.fileName = GenerateOptions.getFileNameFromCall(call);
        this.mediaSize = GenerateOptions.getMediaSizeFromCall(call);
        this.timeout = GenerateOptions.getTimeoutFromCall(call);
    }

    @NonNull
    public String getFileName() {
        return fileName;
    }

    @NonNull
    public PrintAttributes.MediaSize getMediaSize() {
        return mediaSize;
    }

    public long getTimeout() {
        return timeout;
    }

    @NonNull
    private static String getFileNameFromCall(@NonNull PluginCall call) {
        String fileName = call.getString("fileName");
        if (fileName == null) {
            fileName = UUID.randomUUID() + ".pdf";
        }
        return fileName.endsWith(".pdf") ? fileName : fileName + ".pdf";
    }

    @NonNull
    private static PrintAttributes.MediaSize getMediaSizeFromCall(@NonNull PluginCall call) throws Exception {
        PrintAttributes.MediaSize mediaSize;
        switch (call.getString("pageSize", "A4")) {
            case "A3":
                mediaSize = PrintAttributes.MediaSize.ISO_A3;
                break;
            case "A4":
                mediaSize = PrintAttributes.MediaSize.ISO_A4;
                break;
            case "A5":
                mediaSize = PrintAttributes.MediaSize.ISO_A5;
                break;
            case "LETTER":
                mediaSize = PrintAttributes.MediaSize.NA_LETTER;
                break;
            default:
                throw CustomExceptions.PAGE_SIZE_INVALID;
        }
        switch (call.getString("orientation", "PORTRAIT")) {
            case "LANDSCAPE":
                return mediaSize.asLandscape();
            case "PORTRAIT":
                return mediaSize.asPortrait();
            default:
                throw CustomExceptions.ORIENTATION_INVALID;
        }
    }

    private static long getTimeoutFromCall(@NonNull PluginCall call) {
        return call.getInt("timeout", 30000);
    }
}

File: packages/pdf-generator/ios/Plugin/Classes/Options/GenerateOptions.swift
import Foundation
import Capacitor

@objc public class GenerateOptions: NSObject {
    let fileName: String
    let paperSize: CGSize
    let timeout: Int

    init(_ call: CAPPluginCall) throws {
        self.fileName = GenerateOptions.getFileNameFromCall(call)
        self.paperSize = try GenerateOptions.getPaperSizeFromCall(call)
        self.timeout = GenerateOptions.getTimeoutFromCall(call)
    }

    private static func getFileNameFromCall(_ call: CAPPluginCall) -> String {
        let fileName = call.getString("fileName") ?? UUID().uuidString + ".pdf"
        return fileName.hasSuffix(".pdf") ? fileName : fileName + ".pdf"
    }

    private static func getPaperSizeFromCall(_ call: CAPPluginCall) throws -> CGSize {
        let pageSize: CGSize
        switch call.getString("pageSize") ?? "A4" {
        case "A3":
            pageSize = CGSize(width: 842, height: 1191)
        case "A4":
            pageSize = CGSize(width: 595, height: 842)
        case "A5":
            pageSize = CGSize(width: 420, height: 595)
        case "LETTER":
            pageSize = CGSize(width: 612, height: 792)
        default:
            throw CustomError.pageSizeInvalid
        }
        switch call.getString("orientation") ?? "PORTRAIT" {
        case "LANDSCAPE":
            return CGSize(width: pageSize.height, height: pageSize.width)
        case "PORTRAIT":
            return pageSize
        default:
            throw CustomError.orientationInvalid
        }
    }

    private static func getTimeoutFromCall(_ call: CAPPluginCall) -> Int {
        return call.getInt("timeout") ?? 30000
    }
}
